### ISSUE TITLE
Notepad++ exclude folder find in files filter

### ISSUE THREAD

**Issue by evandrocoan** [notepad-plus-plus/notepad-plus-plus#2433]
### Notepad++ exclude folder on the `Find in Files` filter

Originally posted on: http://stackoverflow.com/questions/4198829/notepad-find-in-files-filter-exclude

> I'm looking for a way to get n++'s find in files dialog to exclude certain files...
> 
> the only reference I've found so far is this http://sourceforge.net/project/shownotes.php?release_id=536795&group_id=189927 , but, unfortunately, it doesn't work. 
> 
> I'm using *[^*.dll] to, I think, search for everything BUT dll files.  
> 
> How can I resolve this ?

Is there how to exclude folders from the find in files dialog? I need to exclude the `.git` folders from the search, as the are very big and there are several of them.

![image](https://cloud.githubusercontent.com/assets/5332158/19440114/659f0d58-945f-11e6-9f2d-070851500037.png)

**Comment by dail8859** [comment]
You can't. It doesn't work of real regular expressions. It uses the windows API and only supports wildcards (`*`) and I think question marks.

To search in multiple file types use something like `*.cpp *.h *.xml`

**Comment by nikoudel** [comment]
Regex in Filters could become one of the best features of Notepad++ :(

**Comment by robsiera** [comment]
At least every two years I find myself investigating if NotePad++ already has the long wanted "Exclude files filter".  I'm not complaining though. Keep up the good work!

**Comment by Spin45** [comment]
Please add the ability to exclude folders/files in searches.

**Comment by Nightblade** [comment]
Just checking, this isn't fixed yet is it?

**Comment by PhilipOakley** [comment]
The solution from https://notepad-plus-plus.org/community/topic/10653/exclude-directories-from-the-search/5 is to temporarily set those folders you want to exclude to be 'hidden' folders (hiding just the top level sub-dir is sufficient), and this then stops Notepad++ descending into that sub-directory.

Worked for me just now!

**Comment by Nightblade** [comment]
@PhilipOakley Thanks, but technically that's a workaround isn't it?  ;)

**Comment by sasumner** [comment]
@Nightblade Sometimes...and sometimes sadly...workarounds are all we get. :)

**Comment by PhilipOakley** [comment]
@Nightblade Correct, it's just a workaround, until such time as it is included in the (searchable) documentation :wink:

**Comment by sudhir600** [comment]
Here is a easy solution -
go to your project and find `.git` or whatever folder you want (e.g `node_module`) and right click then `properties`.

check as hidden (Apply changes to this folder, subfolders and files).

![exclude_folder_from_search_in_notepad++](https://user-images.githubusercontent.com/4952560/58074906-fc1b5680-7bc3-11e9-8d54-4661339d4371.png)

_and then search on notepad++ (do not check "`In Hidden Folders`")_

**Comment by nico008** [comment]
Thank you for this workaround but it would be good to have a long term solution for this.

**Comment by Fenek912** [comment]
Still waiting 😜

**Comment by evandrocoan** [comment]
No need to keep waiting for it. You can implement it yourself. If you do not know how to program computers or on C++ specifically, this is a good opportunity for you to learn. However, if you do not have time, you can just keep waiting for it, or hire someone to do it. If you do not have enough money to hire someone, you can save money until you get enough. If you cannot save money, then you have more important things to do and that is the end. I personally, stopped using Notepad++ 3 years ago. Nowadays I use Sublime Text, but I am thinking to migrate to VSCode, if it is not too much slow: 
1. [https://forum.sublimetext.com#41375](https://forum.sublimetext.com/t/sublime-text-versus-visual-studio-code-in-2019/41375)  Sublime Text versus Visual Studio Code in 2019

**Comment by nanw1103** [comment]
I love notepad++ because it's fast and straightforward. However, lacking the feature of excluding certain folder patterns is a key blocker for we coders: many code repositories have 3rd party stuff in the folder. Searching the repository but without the capability to exclude them has a major performance impact and makes the "search in files include subfolder" unusable most of the time.

**Comment by donho** [comment]
It's implemented in v7.8.2:
![image](https://user-images.githubusercontent.com/90293/70520844-78648f80-1b3e-11ea-82ce-1d0189a052c9.png)

**Comment by donho** [comment]
Just an idea, but nothing is promised.
Do you guys think the following syntax is feasible?
`*.* **-F**folderName2Exclude1 **-F**folderName2Exclude2`

**Comment by xylographe** [comment]
Most useful to me would be a list of folder name globs (i.e. wildcard support) somewhere in the Preferences dialogue. That would allow me to always exclude folders named _bin_ or _obj_, and folders whose names start with a dot: **`.*;bin;obj`**.

Because folder names with spaces are not uncommon, I would suggest the semi-colon as the separator between globs.

**Comment by donho** [comment]
> Most useful to me would be a list of folder name globs (i.e. wildcard support) somewhere in the Preferences dialogue. 

I don't think it's a good idea:
1. Adding more items in preference dialog make it more complex.
2. If user puts a folder name by accident and forget to remove it, the folder will always excluded and he/she will never understand why.

I love KISS, and I believe keep Find in files filter settings in the Find in file is the best way. Your filter will be memorized anyway so why bother to make things more complicated?

**Comment by donho** [comment]
> Because folder names with spaces are not uncommon

Good point.

> I would suggest the semi-colon as the separator between globs.

But since space has been used as separator, and we continue using it after the exclusion of files, I would keep it. Here is suggested solution:

`*.* -FfolderName2Exclude1 -F"folder name with space to exclude"`

**Comment by xylographe** [comment]
I agree, space separator and double quotes is better than `;` separator.

To me the location (Preferences or Find dialogue) is not important, I can see why you would prefer the Find dialogue. But, I would prefer to keep the current file filter as is, i.e. file name globs _only_, and add a new folder-exclude field to the Find dialogue, because filtering on file type depends very much on the type of project (e.g. `*.cs !*.Designer.cs` makes sense for a CSharp source code project, while `*.srt *.ssa *.ass *.xml` could be useful for a collection of subtitles), while folder exclusion is not project specific (e.g. `.*` folders are typically used for bookkeeping purposes by auxiliary programs).

**Comment by robsiera** [comment]
>  add a new folder-exclude field to the Find dialogue
I second that!

**Comment by donho** [comment]
Just checked the code. It's possible to implement it, but I'm not sure it's what users want:
If an user try to search in `c:/myProj/` and he/she excludes folder "logs", then not only `c:/myProj/logs/` will be excluded, but also `c:/myProj/src/logs/`and its sub-folders will be ignored.
Is it the expected behaviour?

**Comment by evandrocoan** [comment]
> Is it the expected behavior?

Depending on the configuration. If I exclude like `-log/*;` then yes. It should exclude both `c:/myProj/logs/` and `c:/myProj/src/logs/`

1. But if I exclude it like `-log/;` then only `c:/myProj/logs/` dir should be excluded.
1. But if I exclude it like `-log/file.txt;` then only `c:/myProj/logs/file.txt` file should be excluded.

**Comment by xylographe** [comment]
> Is it the expected behaviour?

Yes, that is exactly what I expect. All folders named _logs_ shall be excluded: in the top-folder, any sub-folder, any sub-sub-folder, etc.

I do **not** expect paths (e.g. `src/tmp`) to be supported, as this would become too complicated:
1. `/src/*/tmp` excludes `c:/myProj/src/ABC/tmp/` and `c:/myProj/src/XYZ/tmp/` (`c:/myProj/` is the root of the search-folder-hierarchy).
2. `src/*/tmp` also excludes `c:/myProj/ALT1/src/ABC/tmp/` and `c:/myProj/ALT2/src/ABC/tmp/` (can match anywhere in the search-folder-hierarchy, because `src/*/tmp` is un-rooted).
3. `/src/**/tmp` excludes `c:/myProj/src/AA/BB/tmp/` and `c:/myProj/src/AA/CC/tmp/` (in addition to the folders excluded by [1]).
4. `src/**/tmp` excludes `c:/myProj/ALT3/SUB2/src/XX/YY/ZZ/tmp/` and `c:/myProj/ALT7/LL/NN/OO/tmp/` (in addition to the folders excluded by [2] and [3]).

None of those are necessary.

But, I do expect `tmp-*` to exclude all folders (at any level) whose names start with _tmp-_.

To summarize:
- Only folder name globs are allowed (no paths).
- Globs are matched against the name of each encountered folder.
- When a folder name matches a glob that folder will be completely skipped (as if the folder had been marked _hidden_).

**Comment by donho** [comment]
> But if I exclude it like -log/file.txt; then only c:/myProj/logs/file.txt file should be excluded.

This feature won't be implemented for sure due to its complesity.

> I do not expect paths (e.g. src/tmp) to be supported, as this would become too complicated:

Yes, you're right - they won't be supported.

> But, I do expect tmp-* to exclude all folders (at any level) whose names start with tmp-.

Not sure if it's possible. I'll check that.

**Comment by xylographe** [comment]
> Not sure if it's possible. I'll check that.

I believe, [PathMatchSpec](https://docs.microsoft.com/en-us/windows/win32/api/shlwapi/nf-shlwapi-pathmatchspecw) supports `*` and `?` wildcards. If `pszFile` contains the folder name, it should work the same as with a file name.

**Comment by donho** [comment]
Already in v7.8.2 now.
dc48cee5f7f4efe1f70e7269fe5e46ac8b861b1b

**Comment by xylographe** [comment]
@donho 
Excluding **files** based on file name has been added in commit dc48cee5f7f4efe1f70e7269fe5e46ac8b861b1b.
But, excluding **folders** based on folder name (as discussed above) is still open.

**Comment by PhilipOakley** [comment]
Hi, back again. I'm still looking for directory exclusion. 

In may case it's the `vcpkg` directory and all its vast contents which takes forever to search (and sometimes fails if there are too many search matches). This is for the visual studio build of the Git-for-Windows repository which needs many many support packages to compile Git and shim in the linux-Windows compatibility layers - these are all in the vcpkg sub-directory!.

In may case a simple `!dirs\dirs` format is sufficient, with a minimum of one `\` directory separator in the string (as a way of indicating this is a 'path' (rather than file) exclusion (it does allow a qualified path via multi-layered path e.g. a\b\c). I'm happy without any wild cards. Using `-` rather than `!` would also work.

**Comment by PhilipOakley** [comment]
A quick bit of retrospective search found https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/FileBrowser/fileBrowser.cpp#L946-L949 as the likely place to sneak in the termination of a recursive search down the directories, based on the current "suggestion" of using isHidden  - just maybe.

**Comment by Flamefire** [comment]
Can we get an additional edit field in the search dialog which is called `exclude` and where you can put regular globs for files and folders to exclude? The most prominent would be `build`: Do not search the build folder. Especially for (largish) CMake-generated VS solutions this is vital or the search is almost unusable.
It could even be hidden by default (when empty) and you have to expand to use it.

See e.g. VSCode which has a very nice solution to this.

**Comment by ArkadiuszMichalski** [comment]
@donho 
Why close? Excluding folder is still missing. Just today I also felt the lack of this possibility, hidding the folder via attribute is not convenient. Any form that excludes folders (even without *?) Would be better than no one.

**Comment by dail8859** [comment]
@ArkadiuszMichalski It is closed because the feature was implemented. If there are problems with the functionality, please open a new issue. If you are unsure how to exclude files you can either look at the commit details, Notepad++ release notes, post on the community forum, or consult the N++ manual.

**Comment by Flamefire** [comment]
Can you point to the commit? Because as noted above:

> Excluding files based on file name has been added in commit dc48cee.
> But, excluding folders based on folder name (as discussed above) is still open.

**Comment by ArkadiuszMichalski** [comment]
@dail8859 Bug was opened for excluding folders:
- title: "Notepad++ exclude folder find in files filter "
- description: "Is there how to exclude folders from the find in files dialog? I need to exclude the .git folders from the search, as the are very big and there are several of them."

Implemented future in https://github.com/notepad-plus-plus/notepad-plus-plus/commit/dc48cee5f7f4efe1f70e7269fe5e46ac8b861b1b add filter only for files. That's not enough, and I can see that not only for me. Even solution sugested by @donho in https://github.com/notepad-plus-plus/notepad-plus-plus/issues/2433#issuecomment-564062343 is better than nothing.

I will not open a new bug because here we have some discussion with possible solutions.

**Comment by donho** [comment]
True, this issue is for excluding folders.

**Comment by dail8859** [comment]
@ArkadiuszMichalski Ah yes. Misunderstanding on my part. Apologies for that.

**Comment by Neustradamus** [comment]
Currently, it is possible to exclude filenames or file extentions in command line too?

I have created a ticket about command line:
- https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8350

**Comment by apalopohapa** [comment]
FWIW, I also run into the need to exclude folders by name to avoid brutally long search times. Setting them as hidden is not practical. Filter by filename does not apply (they all have files of the same type).

**Comment by Cerno-b** [comment]
I'd like to mention how Beyond Compare does this very nicely. I'm a big fan of their solution to file/directory filtering.

For them it is important to treat directories as a work item, which will not be relevant to Notepad++ since you only work directly on files. Therefore I'm going to ignore Beyond Compare's directory filtering here, which only includes/excludes directories themselves and does not care for their contents.

Beyond Compare's file filtering, however, allows for including or excluding files that reside in specific directories. What makes this beautiful to use is that your directory filter must always include the file filters, so you don't have to worry specifying a separate syntax for filtering a file or a folder.

So instead of saying give me `*.txt` but not in directory `c:\temp`, you are saying give me `*.txt` but ignore `c:\temp\*.*`.

They also provide a visual editor...

![image](https://user-images.githubusercontent.com/16470072/118126052-0c9fe980-b3f8-11eb-93ee-5f7e52ff9cde.png)

...which translates into a single string (so the dialog is optional):

`...\look_here\...\*.txt;-notme.txt;-...\do_not_look_here\...\*.txt`

The ellipsis (`...`) is a wildcard for one or multiple directories. 

So assuming `C:\` is your working directory, the filter above will include 
* `C:\temp\look_here\sub1\sub2\sub3\file1.txt`
* `C:\temp\look_here\sub1\sub2\file2.txt`

It will ignore 
* `C:\temp\look_here\do_not_look_here\sub2\sub3\file.txt` 
* `C:\temp\look_here\sub1\sub2\sub3\notme.txt`

You can also use wildcards in your directory names:
`...\look_*\...\*.txt`

This will give you
* `C:\temp\look_here\sub1\sub2\sub3\file1.txt`
* `C:\temp\look_there\sub1\sub2\file2.txt`

Of course the syntax will have to be adapted to Notepad++'s existing syntax, so I guess it will be `!` instead of `-` and space instead of semicolon. You would still need to distinguish between the name wildcard `*` and the subtree wildcard `...`, otherwise you won't be able to distinguish these two cases:

* `test\...\go.txt`
* `test\*\go.txt`

The first will accept 
* `test\a\go.txt`
* `test\a\b\c\go.txt`

The second will accept
* `test\a\go.txt`
but not
* `test\a\b\c\go.txt`

So in the end, the `...` wildcard is optional if you want to force the user to determine the exact structure of the subdirectories, but I find it very convenient to have a way to say "get me everything from `x`, but don't look into `temp` directories anywhere along the path".

**Comment by timint** [comment]
+1 For a file mask filter for excluded paths in the search dialog. This is how Visual Studio Code does it:

![image](https://user-images.githubusercontent.com/359192/138203848-386d77b5-398a-4171-b493-0346b2954612.png)

This is where I find it convenient to put it:
![image](https://user-images.githubusercontent.com/359192/138204156-62541634-0bf5-4d36-aad4-827448fa5607.png)

**Comment by donho** [comment]
A "Exclude" combobox will be added, that allows users to provide the list of excluding folder names with the same format as "Filters" - allow `*` and use white space as separator. The provided name(s) will be used for excluding only the 1st level of matched directory. If users need to exclude folders with the same name (or matched the specific pattern) in all levels, a prefix `+` should be added. 

Here is a sample:
Filters: `*.*`
Directory: `c:\myProject\`
Exclude: `bin +log* +test`

So the excluded directories could be:
1. *c:\myProject\bin\\*
2. *c:\myProject\log\\*
3. *c:\myProject\logs\\*
4. *c:\myProject\src\log4j\\*
5. *c:\myProject\test\\*
6. *c:\myProject\src\test\\*

The following directories will not be excluded:
1. *c:\myProject\foo\tests\\*
2. *c:\myProject\foo\bin\\*

What do you guys think?

**Comment by timint** [comment]
Hey man, I am very much delighted to hear you are looking into this.

At first I was about to propose the double globstar. But then I realized that's gotta be what the + is for, yeah?

I think your proposal is good. I am a little uncertain if not double globstar (meaning directory separator / matching) is the better way to go? It is pretty well known from other frameworks/apps/functions.

If there is a technical limitation why it can't be used then it has to be +.

Hope I got this right:
* = [^\]+
** = .*

**Comment by donho** [comment]
@timint 
I don't know about the notation of double star, and it doesn't seem to me a standard.
If your double star works in filters field, it might work in Exclude field.
Matching file feature has been done (in Filters field) since long time ago, and it won't be changed.
Here we talking only about exclude folder(s) feature.

**Comment by timint** [comment]
I am **only** talking about the exclude filters.

I didn't say it was a standard. I was saying it is pretty well known. It's even implemented in some linux shells. A search on google for double glob star should give enough hits to keep a reader busy for an eternity.

**Comment by donho** [comment]
For people who want to test this new feature, here are the binaries for both [x86 ](https://ci.appveyor.com/api/buildjobs/w3h9apilp3tpmrh8/artifacts/Notepad%2B%2B.Win32.Release.exe)& [x64](https://ci.appveyor.com/api/buildjobs/c1d81evb9o54sms8/artifacts/Notepad%2B%2B.x64.Release.exe).

Please let me know if it works as expected.

**Comment by Neustradamus** [comment]
@donho: First, thanks for your improvements!

But in your example:

```
Here is a sample:
Filters: *.*
Directory: c:\myProject\
Exclude: bin +log* +test

So the excluded directories could be:
c:\myProject\bin\
c:\myProject\log\
c:\myProject\logs\
c:\myProject\src\log4j\
c:\myProject\test\
c:\myProject\src\test\
```

"+test" excludes:
```
- c:\myProject\test\
- c:\myProject\src\test\
```

How it is possible to exclude only c:\myProject\test\ and not c:\myProject\src\test\?

or "test" excludes only c:\myProject\test\?

Example 1:
```
Filters: *.*
Directory: c:\myProject\
Exclude: bin log* test

It will exclude?
c:\myProject\bin\
c:\myProject\log\
c:\myProject\logs\
c:\myProject\test\
```

Example 2 with a subfolder:
```
Filters: *.*
Directory: c:\myProject\
Exclude: bin log* src\test\

It will exclude?
c:\myProject\bin\
c:\myProject\log\
c:\myProject\logs\
c:\myProject\src\test\
```

How to exclude files?

In more, I have opened a ticket here: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8350

**Comment by donho** [comment]
> How it is possible to exclude only c:\myProject\test\ and not c:\myProject\src\test?

Set to `test` will exclude only the first level (c:\myProject\test\) but not c:\myProject\src\test\

> or "test" excludes only c:\myProject\test?

Yes

> How to exclude files?

Use Filters field (Mouse hovering the "Filters" label will get the hint).

**Comment by Neustradamus** [comment]
> > How it is possible to exclude only c:\myProject\test\ and not c:\myProject\src\test?
> 
> Set to `test` will exclude only the first level (c:\myProject\test) but not c:\myProject\src\test\
> 
> > or "test" excludes only c:\myProject\test?
> 
> Yes

Nice
> 
> > How to exclude files?
> 
> Use Filters field (Mouse hovering the "Filters" label will get the hint).

OK

And about subfolders?
Example "c:\myProject\src\test\"

Please look folders, subfolders, sub-subfolders here:
- https://github.com/notepad-plus-plus/notepad-plus-plus/issues/8350

**Comment by PhilipOakley** [comment]
> Please let me know if it works as expected.

For my use case of hiding the vast `vcpkg` sub-folder in the Git-for Windows compilable checkout, that exclude has worked fine. It was on a very simple test, but it is a worthwhile valuable first step (hopefully the other capabilities within the exclude syntax covers the other cases).

**Comment by donho** [comment]
@timint 
I've checked double-asterisk (**) wildcards, and indeed it's interesting standard (it's in bash anyway).
However, it's hard to parse (for the current implementation) the pattern if double globstar is included. 
Thank you anyway for your suggestion.

**Comment by donho** [comment]
@Nightblade 
> And about subfolders?
Example "c:\myProject\src\test"

I am implementing the [2nd version](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10924#issuecomment-997389496) which uses Filters field only (No more Exclude field). I will publish the usage and binaries link here.

**Comment by timint** [comment]
@donho Thank you for considering. I use to say; always know what options you are turning down when making choices. 🙂

I am fine with whatever you come up with. Will your implementation support file masks such as excluding *.log files?

**Comment by donho** [comment]
Here is the new implementation: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/10927
In this implementation, the Filter field is used for putting the list of folder patterns to be excluded (instead of using a new field).

**Comment by mere-human** [comment]
The syntax with `+` looks very unusual. How about to use a syntax similar to `.gitignore`?
- use `,` or `;` to split entries
- `!` for negation (exclude)
- `**\test` to exclude every leaf `test` directory (so called first level directories)

See this for more details: https://git-scm.com/docs/gitignore

**Comment by PhilipOakley** [comment]
> * `**\test` to exclude every leaf `test` directory (so called first level directories)

The .gitignore style can be confusing to some because of the confusion between leading and embedded dir separators and a trailing dir separator which indicate different things, along with confusions about the 'root' of the ignore pattern (less of a problem here as the 'root' for the pattern well defined, but is mobile for gitignore).

**Comment by donho** [comment]
@mere-human 
> The syntax with + looks very unusual. 

Which symbol more meaningful in your opinion then? 

> How about to use a syntax similar to .gitignore?

The idea is not redo the syntax of Filters pattern, but add exclusion of folder into it.

**Comment by timint** [comment]
Looking at the Notepad++ documentation I understand the current syntax used for search in files today is: `*.*`, `name.*`, and negative matching is `!*.ext`.
https://npp-user-manual.org/docs/searching/

Those are already fully compatible with the double globstar system used by Linux Shells, Gitignore, Visual Studio Code, Grunt, Gulp, and many other. From my understanding it is only the introcution of + that makes it incompatible? It shouldn't be too hard making + an alias for ** or vice versa to support double globstar in the future. Maybe that should be a feature request by itself?

Here is a decent article describing how the syntax operates:
https://www.atlassian.com/git/tutorials/saving-changes/gitignore

A directory separator would indicate when it's a directory in the rule. E.g. `**/logs/*.log` or `logs/*` so there is no need for new identifiers.

**Comment by donho** [comment]
@timint 
The reason of not making pattern syntax as `!**/log*` instead of `!+\log*` is prevent from misleading users. If double globstar were used, that could give users an illusion that `**/logs/debug.log` or `logs/**/debug.log` patterns work also, then bug report coming...
Currently find in files searching feature doesn't contain a shell interpreter to make above patterns work, and it's hard to make it compatible also in the future.
Therefore I think it's better to create a simple, comprehensive and non-ambiguous system to get job done.

**Comment by Neustradamus** [comment]
It is included in Notepad++ 8.2.0:
- https://notepad-plus-plus.org/news/v82-released/
- https://github.com/notepad-plus-plus/notepad-plus-plus/releases/tag/v8.2
